cpp/libsarmata-client/sarmata_client.cpp

```diff
@@ -0,0 +1,228 @@
+            key_value.push_back(field);
+        }
+
+        if (key_value.size() == 2)
+        {
+            auto field = recognition_config.add_additional_settings();
+            field->set_key(key_value[0]);
+            field->set_value(key_value[1]);
+            std::cout << "key: " << field->key() << " | value: " << field->value() << std::endl;
+        }
+        else
+        {
+            std::cout << "Skipping invalid session settings line: " << line << std::endl;
+        }
+    }
+}
+
+void build_recognition_config(const SarmataClientConfig& config, unsigned int sample_rate_hertz, RecognitionConfig& recognition_config) {
+    recognition_config.set_max_alternatives(config.max_alternatives);
+    recognition_config.set_sample_rate_hertz(sample_rate_hertz);
+    recognition_config.set_no_match_threshold(config.no_match_threshold);
+
+    if (not config.grammar_name.empty()) {
+        recognition_config.set_grammar_name(config.grammar_name);
+    }
+    else {
+        recognition_config.set_grammar_data(config.grammar_data);
+    }
+
+    auto timeouts = recognition_config.mutable_timeout_settings();
+    timeouts->set_no_input_timeout(config.no_input_timeout);
+    timeouts->set_recognition_timeout(config.recognition_timeout);
+    timeouts->set_speech_complete_timeout(config.speech_complete_timeout);
+    timeouts->set_speech_incomplete_timeout(config.speech_incomplete_timeout);
+
+    if (not config.service_settings.empty()) {
+        read_service_settings_option(config, recognition_config);
+    }
+}
+
+std::vector<RecognizeRequest> build_request(const SarmataClientConfig& config, unsigned int audio_sample_rate_hz, const std::string& audio_byte_content)
+{
+    RecognizeRequest request;
+    build_recognition_config(config, audio_sample_rate_hz, *request.mutable_config());
+
+    std::vector<RecognizeRequest> requests;
+    requests.push_back(request);
+
+    unsigned int frame_length = 20;//milliseconds [ms]
+    unsigned int frame_size = frame_length * audio_sample_rate_hz / 1000;//samples
+    for (auto i = 0; i < audio_byte_content.length(); i += frame_size)
+    {
+        RecognizeRequest request;
+
+        request.set_audio_content(audio_byte_content.substr(i,frame_size));
+        requests.push_back(request);
+    }
+
+    return requests;
+}
+
+bool error_response(const RecognizeResponse& response) {
+    const auto status = response.status();
+    const auto is_error = (status == GRAMMAR_LOAD_FAILURE
+                        || status == GRAMMAR_COMPILATION_FAILURE
+                        || status == RECOGNIZER_ERROR
+                        || status == SEMANTICS_FAILURE);
+
+    if (is_error) {
+        const auto& error = response.error();
+        std::cerr << "Received error response: (" << ResponseStatus_Name(status) << ") " << error << std::endl;
+    }
+
+    return is_error;
+}
+
+std::string grpc_status_to_string(const grpc::Status& status) {
```

Comment:
I think that extracting GRPC-related functions (`build_recognition_config`, `build_request`, `error_response` and `grpc_status_to_string`) to separate source file would make `sarmata_client.cpp` easier to read

Reply:
I don't agree that a separate file would improve readability here. What I can offer instead is to move public methods (`DefineGrammar` and `Recognize`) to the top of the file.